src/AMDGPUOffload.cpp

```diff
@@ -0,0 +1,639 @@
+            Expr code_size = Call::make(Int(32), Call::buffer_get_extent, { code_buf_var, 0 }, Call::Extern);
+            Expr code_ptr = Call::make(Handle(), Call::buffer_get_host, { code_buf_var }, Call::Extern);
+            Stmt init_kernels = call_extern_and_assert("halide_amdgpu_initialize_kernels",
+                                                       { module_state_ptr(), code_ptr, cast<uint64_t>(code_size), runtime_ptr, cast<uint64_t>(runtime_size) });
+            s = Block::make(init_kernels, s);
+        }
+
+        // TODO: This can probably go away due to general debug info at the submodule compile level.
+        debug(1) << "Amdgpu device code module: " << device_code << "\n";
+
+        return s;
+    }
+};
+
+}  // namespace
+
+Stmt inject_amdgpu_rpc(Stmt s, const Target &host_target,
+                        Module &containing_module) {
+    Target target(Target::Linux, Target::X86, 64);
+    Module amdgpu_module(pipeline_module_name, target.with_feature(Target::NoRuntime));
+    InjectAmdgpuRpc injector(amdgpu_module);
+    s = injector.inject(s);
+
+    if (!amdgpu_module.functions().empty()) {
+        containing_module.append(amdgpu_module);
+    }
+
+    return s;
+}
+
+Buffer<uint8_t> compile_module_to_amdgpu_shared_object(const Module &device_code) {
+    llvm::LLVMContext context;
+    std::unique_ptr<llvm::Module> llvm_module(compile_module_to_llvm_module(device_code, context));
+
+    // Write intermediate bitcode to disk if requested.
+    // TODO: We really need something better than this. This won't
+    // work in non-trivial JIT or AOT programs.
+    std::string bitcode_dump_path = get_env_variable("HL_AMDGPU_DUMP_BITCODE");
+    if (!bitcode_dump_path.empty()) {
+        auto fd_ostream = make_raw_fd_ostream(bitcode_dump_path);
+        compile_llvm_module_to_llvm_bitcode(*llvm_module, *fd_ostream);
+        debug(0) << "Wrote Amdgpu device bitcode to " << bitcode_dump_path;
+    }
+
+    llvm::SmallVector<char, 4096> object;
+    llvm::raw_svector_ostream object_stream(object);
+    compile_llvm_module_to_object(*llvm_module, object_stream);
+
+    int min_debug_level = device_code.name() == runtime_module_name ? 3 : 2;
+    if (debug::debug_level() >= min_debug_level) {
+        debug(0) << "AMDGPU device code assembly: " << "\n";
+        llvm::SmallString<4096> assembly;
+        llvm::raw_svector_ostream assembly_stream(assembly);
+        compile_llvm_module_to_assembly(*llvm_module, assembly_stream);
+        debug(0) << assembly.c_str() << "\n";
+    }
+
+    auto obj = Elf::Object::parse_object(object.data(), object.size());
+    internal_assert(obj);
+
+    // Generate just one .text section.
+    obj->merge_text_sections();
+
+    // Make .bss a real section.
+    auto bss = obj->find_section(".bss");
+    if (bss != obj->sections_end()) {
+        bss->set_alignment(128);
+        // TODO: We should set the type to SHT_NOBITS
```

Comment:
These comments are basically useless to you, I think.

Reply:
Are they? @dsharletg

Reply:
Yes, at the very least the reference to '8998' is nonsense in this context.